### PR TITLE
Support HashiCorp Vault namespaces (Enterprise feature) and allow non-standard transit engine names

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

Currently the EnvelopeEncryption filter fails to operate against a HCE.  HCEs make use of the Enterprise namespace feature,
which means the URLs generated by the Kms implementation are wrong.

I took the approach of letting the vaultUrl provide full path to the transit engine.  That gives support for both namespaces, and gives scope to work with non-standard engine path names.

An alternative would have been to let the implementation accept a namespaces configuration parameter, with would be passed to the [X-Vault-Namespace](https://developer.hashicorp.com/vault/docs/enterprise/namespaces#vault-api-and-namespaces) on the request.  A separate parameter would be required to allow for the customisation the name of the transit engine.

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [x] Reference relevant issue(s) and close them after merging
- [x] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by k-wall** [comment]
With this fix, the filter works with HCE too.  Here's I'm using a kroxylicious namespace created beneath the parent (admin) in my HCE instance.

```
    - type: EnvelopeEncryption
      config:
        kms: VaultKmsService
        kmsConfig:
          vaultUrl: https://sample-cluster-public-vault-a157adce.14dc13da.z1.hashicorp.cloud:8200/v1/admin/kroxylicious/transit/
          vaultToken: hvs.CAESIA9yak-E_ixxxxxxxxxxxxxxxxxxxxxxxxxxxx
        selector: TemplateKekSelector
        selectorConfig:
          template: "topic.${topicName}"
```

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java [hunk 1]
```diff
@@ -186,4 +207,8 @@ else if (r.statusCode() != 200) {
             return handler.apply(r);
         };
     }
+
+    /* testing */ URI getEngineUri() {
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java:211]: Nit: i think we have `@visbableForTesting` available

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTest.java [hunk 2]
```diff
@@ -50,6 +55,44 @@ void testRequestTimeoutWaitingForHeaders() {
...
+                Arguments.of("basic", uri.resolve("v1/transit"), "/v1/transit/"),
+                Arguments.of("trailing slash", uri.resolve("v1/transit/"), "/v1/transit/"),
+                Arguments.of("single namespace", uri.resolve("v1/ns1/transit"), "/v1/ns1/transit/"),
+                Arguments.of("many namespaces", uri.resolve("v1/ns1/ns2/transit"), "/v1/ns1/ns2/transit/"),
+                Arguments.of("non standard engine name", uri.resolve("v1/mytransit"), "/v1/mytransit/"));
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource
+    void legalEnginePaths(String name, URI uri, String expected) {
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTest.java:72]: If I was to see a test failure for `legalEnginePaths` it wouldn't be clear what was going wrong. Something like `shouldAcceptUrlPath` makes it clear that we expect the arguments to be accepted by our code.

File: kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTest.java [hunk 3]
```diff
@@ -50,6 +55,44 @@ void testRequestTimeoutWaitingForHeaders() {
...
+        var uri = URI.create("https://localhost:1234");
+        return Stream.of(
+                Arguments.of("unrecognized version", uri.resolve("v999/transit")),
+                Arguments.of("missing engine", uri.resolve("v1")),
+                Arguments.of("empty engine", uri.resolve("v1/")));
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource
+    void illegalEnginePaths(String name, URI uri) {
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault/src/test/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsTest.java:90]: Similar to the legalEnginePaths above, I'm not a fan of the name.

N.B the name makes sense for the argument source!

**Review by SamBarker — APPROVED** [review]
LGTM

Give or take naming nits

**Comment by robobario** [comment]
Is it a bit confusing that `vaultUrl` can be `https://vault:8080` or `https://vault:8080/v1/your/path/to/transit`? It means `vaultUrl` in config can be either a base url, or a base transit url. Should we split it into `vaultBaseUrl` and `transitEnginePath` (default "/v1/transit")?

**Review by robobario — APPROVED** [review]
LGTM, thoughts about whether we should break up the config aren't a blocker

**Comment by k-wall** [comment]
@robobario you make a good point.  The vaultUrl field name no longer describes the purpose.  I think I prefer a single field.

How about I rename the field vaultTransitEngineUrl and make the specification of the path mandatory?

**Comment by robobario** [comment]
> How about I rename the field vaultTransitEngineUrl and make the specification of the path mandatory?

Yep that works for me, thanks

**Comment by SamBarker** [comment]
> How about I rename the field vaultTransitEngineUrl and make the specification of the path mandatory?

Yes, assuming we don't need to support any other engines in the future.

**Comment by k-wall** [comment]
> > How about I rename the field vaultTransitEngineUrl and make the specification of the path mandatory?
> 
> Yes, assuming we don't need to support any other engines in the future.

I can't foreesee myself.  Maybe the Kroxylicious core might have a HashiCorp Secrets integration, but I don't think that would be a concern for EE or KMS services.

File: kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java (hunk 1 above)
Comment (k-wall) [reply kroxylicious-kms-provider-hashicorp-vault/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKms.java:211]: The annotation is part of kroxylicious-core. It probably shouldn't be. https://github.com/kroxylicious/kroxylicious/issues/731

I don't want to fix this now.

**Comment by k-wall** [comment]
@strimzi-ci run system tests

**Comment by k-wall** [comment]
@franvila it looks to me like the systests for this PR ran with the wrong production code.  The tests pass locally.
I'll try running them again.

**Comment by k-wall** [comment]
@strimzi-ci run system tests

**Comment by k-wall** [comment]
I'm sorry, the mistake is mine.  

This PR refactors the `vaultUrl` config key to have a different name.  I thought I'd got all the occurrences, but I missed one in the system-test tree.

**Comment by k-wall** [comment]
> I'm sorry, the mistake is mine.
> 
> This PR refactors the `vaultUrl` config key to have a different name. I thought I'd got all the occurrences, but I missed one in the system-test tree.  I have no idea why my EESTs worked locally.

**Comment by k-wall** [comment]
@strimzi-ci run system tests

**Review by franvila — APPROVED** [review]
LGTM

**Pull request merged**